distributed/core.py

```diff
@@ -516,6 +516,7 @@ async def handle_stream(self, comm, extra=None, every_cycle=[]):
                             handler = self.stream_handlers[op]
                             if is_coroutine_function(handler):
                                 self.loop.add_callback(handler, **merge(extra, msg))
+                                await gen.sleep(0)
```

Comment:
For reference, when looping through messages here:

https://github.com/dask/distributed/blob/a6db3a1967b936d8942cd1c7b98bd683f117e37d/distributed/core.py#L508

handlers which are synchronous functions are executed in order, while coroutine handlers are added to the event loop as a callback then executed in reserve order. This causes issues with our task stealing code as we ended up releasing keys before they were stolen.

To avoid executing operations in reverse order I added `gen.sleep(0)` to force the next I/O loop iteration where the most recently added callback is run. I _think_ this is okay.

For my particular use case there's a workaround where I no longer need asynchronous `Scheduler.add_worker`/`.remove_worker` methods. That said, I think the changes here are still an improvement to the current situation

Reply:
We have a few tests in dask-cuda that check the behavior of Device<->Host<->Disk spilling and I noticed after the 2.19 release one of them has broken, I managed to track it down to this line specifically. The test in question happens in https://github.com/rapidsai/dask-cuda/blob/branch-0.15/dask_cuda/tests/test_spill.py#L409-L411, where we assert that the zict dictionaries are empty after deleting `cdf2`, which is the object being spilled. It seems that this is because we're not awaiting for `Worker.delete_data` somewhere, as per the warning below that doesn't happen if I comment `await gen.sleep(0)` out:

```python
dask_cuda/tests/test_spill.py::test_cudf_device_spill[params0]
  /datasets/pentschev/miniconda3/envs/r-102-0.14/lib/python3.7/inspect.py:732: RuntimeWarning: coroutine 'Worker.delete_data' was never awaited
    for modname, module in list(sys.modules.items()):
```

I _think_ that the only place where `Worker.delete_data` would be called and should be awaited is in https://github.com/dask/distributed/blob/4f878b420b349ee725de5ef64fd5e664dedb8aba/distributed/scheduler.py#L2791-L2800, but I don't have anything better than my guess at this time because it's really hard for me to understand all the async black magic. I'm gonna continue trying to figure this out, but any suggestions on how to pinpoint that are appreciated!